src/arm/96boards.c

```diff
@@ -85,6 +96,99 @@ mraa_96boards_pininfo(mraa_board_t* board, int index, int sysfs_pin, char* fmt,
     pininfo->gpio.mux_total = 0;
 }
 
+// DB820c
+static mraa_result_t
+mraa_db820c_mmap_unsetup()
+{
+	if (mmap_reg == NULL) {
+		syslog(LOG_WARNING, "db820c mmap: null register nothing to unsetup");
+		return MRAA_ERROR_INVALID_RESOURCE;
+	}
+	munmap(mmap_reg, mmap_size);
+	mmap_reg = NULL;
+	close(mmap_fd);
+
+	return MRAA_SUCCESS;
+}
+
+mraa_result_t
+mraa_db820c_mmap_write(mraa_gpio_context dev, int value)
+{
+	uint32_t offset = (0x1000 * dev->pin);
+
+	if (value) {
+		*(volatile uint32_t*) (mmap_reg + offset + 0x04) |= (uint32_t) (1 << 1);
+	} else {
+		*(volatile uint32_t*) (mmap_reg + offset + 0x04) &= ~(uint32_t) (1 << 1);
+	}
+
+	return MRAA_SUCCESS;
+}
+
+int
+mraa_db820c_mmap_read(mraa_gpio_context dev)
+{
+	uint32_t value;
+	uint32_t offset = (0x1000 * dev->pin);
+
+	value = *(volatile uint32_t*) (mmap_reg + offset + 0x04);
+	if (value & (uint32_t) (1 << 0)) {
+		return 1;
+	}
+
+	return 0;
+}
+
+mraa_result_t
+mraa_db820c_mmap_setup(mraa_gpio_context dev, mraa_boolean_t en)
+{
+	if (dev == NULL) {
+		syslog(LOG_ERR, "db820c mmap: context not valid");
+		return MRAA_ERROR_INVALID_HANDLE;
+	}
+
+	/* disable mmap if already enabled */
+	if (en == 0) {
+		if (dev->mmap_write == NULL && dev->mmap_read == NULL) {
+			syslog(LOG_ERR, "db820c mmap: can't disable disabled mmap gpio");
+			return MRAA_ERROR_INVALID_PARAMETER;
+		}
+		dev->mmap_write = NULL;
+		dev->mmap_read = NULL;
+		mmap_count--;
+		if (mmap_count == 0) {
+			return mraa_db820c_mmap_unsetup();
+		}
+		return MRAA_SUCCESS;
+	}
+
+	if (dev->mmap_write != NULL && dev->mmap_read != NULL) {
+		syslog(LOG_ERR, "db820c mmap: can't enable enabled mmap gpio");
+		return MRAA_ERROR_INVALID_PARAMETER;
+	}
+
+	if (mmap_reg == NULL) {
+		if ((mmap_fd = open(MMAP_PATH, O_RDWR)) < 0) {
+			syslog(LOG_ERR, "db820c mmap: unable to open /dev/mem");
+			return MRAA_ERROR_INVALID_HANDLE;
+		}
+		mmap_reg = mmap(NULL, mmap_size, PROT_READ | PROT_WRITE, MAP_SHARED, mmap_fd, DB820C_MMAP_REG);
+
+		if (mmap_reg == MAP_FAILED) {
+			syslog(LOG_ERR, "db820c mmap: failed to mmap");
+			mmap_reg = NULL;
+			close(mmap_fd);
+			return MRAA_ERROR_NO_RESOURCES;
+		}
+	}
+	dev->mmap_write = &mraa_db820c_mmap_write;
+	dev->mmap_read = &mraa_db820c_mmap_read;
+	mmap_count++;
+
+	return MRAA_SUCCESS;
+}
+
```

Comment:
For the above code duplication, we can reuse the same mmap functions. But I haven't found the register documentation for this SoC yet. So, I would suggest you to remove mmap functions for now.

Reply:
Removed